drivers: timer: nrf_rtc_timer: Add clearing of previous CC event

nordic-krch opened the pull request:
Added clearing of CC event which may occure due to previous
CC value which was closed to current counter value.

Fixes #30526.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: timer: nrf_rtc_timer: Add clearing of previous CC event

Added clearing of CC event which may occure due to previous
CC value which was closed to current counter value.

Fixed int_mask initialization.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
See https://github.com/zephyrproject-rtos/zephyr/pull/29810#discussion_r532412748

Apparently, event though event is disabled it can still be generated (for previous CC value) and needs to be cleared.
I've moved event enabling before even though it seems to not be needed but according to RTC spec event generation depends on event being enabled. Spec: https://infocenter.nordicsemi.com/topic/ps_nrf52840/rtc.html?cp=4_0_0_5_21_5#concept_eml_dkj_sr

pabigot commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -160,10 +160,10 @@ static void set_absolute_alarm(uint32_t chan, uint32_t abs_val)
 	do {
 		now = counter();
 
+		event_enable(chan);
```
This seems OK.  My concern was that this could be executed with interrupts enabled.  It can't from `z_clock_set_timeout()` because that holds a lock.  It can't for other channels because `z_nrf_rtc_timer_compare_set()` also holds a lock.  

However, while this will be fine, it seems that anybody who uses `z_nrf_rtc_timer_compare_int_lock()` then invokes `z_nrf_rtc_timer_compare_set()` will have their lock cleared prematurely when the set completes.  That seems like a bug which I'll raise elsewhere.

pabigot replied:
No, I see the `key` parameter should avoid that.

nordic-krch committed:
drivers: timer: nrf_rtc_timer: Fix comparison in z_nrf_rtc_timer_get_ticks

Test was failing due to wrong type conversion.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: timer: nrf_rtc_timer: Add test for reseting CC

Added test for reseting CC timeout. CC is reconfigured just beofre
expiring of the first CC. It is expected that first setting will never
expire.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
I've added additional test for resetting CC which happens just before earlier setting. It has shown that if event occurred when event was disabled in `EVTEN` it will expire immediately after enabling this event because of that first event is enabled then cleared (if previous event occurred).
Moreover, I've noticed that bsim behaves differently and busy wait in case of close event must be longer: full tick instead of half tick.

anangl commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -160,20 +161,21 @@ static void set_absolute_alarm(uint32_t chan, uint32_t abs_val)
 		/* Handle case when previous event may generate an event.
-		 * It is handled by setting CC to now (that's furtherst future),
+		 * It is handled by setting CC to now (far in the future),
 		 * in case previous event was set for next tick wait for half
 		 * LF tick and clear event that may have been generated.
```
> Moreover, I've noticed that bsim behaves differently and busy wait in case of close event must be longer: full tick instead of half tick.

Perhaps it would be good to include this information in the above comment. Right now someone might get confused why for BSim half LF tick is 31 microseconds.

nordic-krch commented:
It seems that bsim is still not working #30507 on master and after this fix, too.

nordic-krch committed:
tests: drivers: timer: nrf_rtc_timer: Adapt test to nrf52_bsim

Adapted test to be able to run on nrf52_bsim

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
west.yml: Update nrf_hw_models revision

Updated revision of nrf_hw_models.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
I've found an issue in nrf52_bsim model. Compare interrupts in RTC were lost if they occurred while interrupt was disabled. I've added PR (https://github.com/BabbleSim/ext_NRF52_hw_models/pull/17) that checks on INTENSET if there are pending interrupt and triggers it.

nordic-krch commented:
> Apparently, event though event is disabled it can still be generated (for previous CC value) and needs to be cleared.
I've moved event enabling before even though it seems to not be needed but according to RTC spec event generation depends on event being enabled.

Further testing shown that RTC behaves as in spec. If event is enabled but interrupt is disabled then event will set EVENTS_COMPARE[n]=1 and when interrupt is enabled that will generate nvic interrupt.
It also means that if event and interrupt is disabled then when compare event occurs it is not recorded at all.

carlescufi commented on west.yml:
```diff
@@ -128,9 +130,10 @@ manifest:
       path: modules/debug/mipi-sys-t
       revision: 75e671550ac1acb502f315fe4952514dc73f7bfb
-    - name: nrf_hw_models
+    - name: ext_NRF52_hw_models
```
next time use `repo-path` so that you don't change the name and hence the bot can work

nordic-krch replied:
changed to: https://github.com/zephyrproject-rtos/zephyr/blob/be2a233296f4496a2baf3b2bed2d48a0a840fc4a/west.yml
but now CI is broken, works locally.

nordic-krch commented:
Removing `DNM` as `nrf_hw_models` fix is in master. Updated revision in west.yml.

anangl commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -160,20 +161,23 @@ static void set_absolute_alarm(uint32_t chan, uint32_t abs_val)
 		 */
 		set_comparator(chan, now);
 		if (counter_sub(prev_cc, now) == 1) {
+			/* Apparently BSIM is not syncing on half tick thus
+			 * added full tick waiting.
+			 */
```
This comment is no longer valid.

nordic-krch replied:
removed.

aescolar commented:
@nordic-krch , when I merge this PR with #30601, I seem to see the same failure that this PR is trying to fix in the new testcase that PR is adding:
```
d_00: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v2.4.0-2276-gfc42e67439a6  ***
d_00: @00:00:00.000000  Running test suite native_cpu_hold
d_00: @00:00:00.000000  ===================================================================
d_00: @00:00:00.000000  START - test_cpu_hold_basic
d_00: @00:00:00.009979  ASSERTION FAIL [chan && chan < (0 + 1)] @ /persistant/scratch/alpi/Zephyr/west_zephyr/zephyr/drivers/timer/nrf_rtc_timer.c:216
d_00: @00:00:00.009979  @ /persistant/scratch/alpi/Zephyr/west_zephyr/zephyr/lib/os/assert.c:45
d_00: @00:00:00.009979  E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
d_00: @00:00:00.009979  E: Current thread: 0x56629c60 (ztest_thread)
d_00: @00:00:00.009979  E: Halting system
```

nordic-krch commented:
@aescolar confirmed same on my side. Investingating...

aescolar commented:
@nordic-krch seems that the cause of the assert is CONFIG_TICKLESS_KERNEL by the test

nordic-krch commented:
@aescolar that was false alarm. Forgot that `nrf52_bsim` required `CONFIG_SYS_CLOCK_TICKS_PER_SEC=32768` option to match timings in the test. That was added in `testcase.yaml` and not in `prj.conf`. I've added it to `prj.conf` now since that makes no harm for hw platforms and makes test work for bsim out of the box.

nordic-krch committed:
drivers: timer: nrf_rtc_timer: Fix non tickless sys clock handling

When tickless mode was disable, sys clock timeout handler was calling
public API function for setting new compare value. Public API function
asserts when chan 0 is used which is reserved for system clock.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@carlescufi can we merge this?

aescolar commented:
> @nordic-krch , when I merge this PR with #30601, I seem to see the same failure in nrf52_bsim that this PR is trying to fix in the new testcase that PR is adding:

Just for the record, this issue is fixed with the latest commit form @nordic-krch
